File: EXPERIMENT 1.1.1/Problem1-SlidingWindow.cpp
#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

bool containsNearbyDuplicate(vector<int>& nums, int k) {
    unordered_set<int> s;

    for (int i = 0; i < nums.size(); i++) {
        if (i > k) {
            s.erase(nums[i - k - 1]);
        }

        if (s.count(nums[i])) {
            return true;
        }

        s.insert(nums[i]);
    }

    return false;
}

int main() {
    int n;
    cin >> n;

    vector<int> nums(n);

    for (int i = 0; i < n; i++) {
        cin >> nums[i];
    }

    int k;
    cin >> k;

    if (containsNearbyDuplicate(nums, k))
        cout << "true";
    else
        cout << "false";

    return 0;
}

File: EXPERIMENT 1.1.1/Problem1-SlidingWindow.py
def contains_nearby_duplicate(nums, k):
    s = set()

    for i in range(len(nums)):
        if i > k:
            s.remove(nums[i - k - 1])

        if nums[i] in s:
            return True

        s.add(nums[i])

    return False


n = int(input())
nums = list(map(int, input().split()))
k = int(input())

print("true" if contains_nearby_duplicate(nums, k) else "false")